### PR TITLE
refactor: Rework core.Spans

### LINKED ISSUE #3209: Rework core.Spans
Broken out of #3038

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #3209

## Description

Simplifies core.Spans a little bit, before removing some old hacky code allowing a nicer way of handling headstore keys within the planner/fetcher system.

Is probably easier to review this commit by commit, the changes are documented by the commit messages.

### CODE REVIEW

File: tests/integration/explain/default/delete_test.go [hunk 1]
```diff
@@ -292,7 +292,12 @@ func TestDefaultExplainMutationRequestWithDeleteUsingNoIds(t *testing.T) {
 							"collectionID":   "3",
 							"collectionName": "Author",
 							"filter":         nil,
-							"spans":          []dataMap{},
+							"spans": []dataMap{
+								{
+									"end":   "/4",
+									"start": "/3",
+								},
+							},
```
Comment (shahzadlone) [inline tests/integration/explain/default/delete_test.go:300]: question: So this was the bug that was caused due to the way `HasValue` was being used?

File: internal/db/fetcher/fetcher.go [hunk 2]
```diff
@@ -278,11 +278,20 @@ func (df *DocumentFetcher) start(ctx context.Context, spans []core.Span, withDel
...
+				// DocumentFetcher only ever recieves document keys
+				//nolint:forcetypeassert
+				valueSpans[i] = core.NewSpan(
+					span.Start.(keys.DataStoreKey).WithDeletedFlag(),
+					span.End.(keys.DataStoreKey).WithDeletedFlag(),
+				)
 			} else {
-				valueSpans[i] = core.NewSpan(span.Start.WithValueFlag(), span.End.WithValueFlag())
+				// DocumentFetcher only ever recieves document keys
+				//nolint:forcetypeassert
```
Comment (shahzadlone) [inline internal/db/fetcher/fetcher.go:290]: suggestion: Have a must DataStoreKey function for the casting? and do this lint ignore there once for all these 4 casts?

File: internal/keys/datastore_doc.go [hunk 3]
```diff
@@ -41,7 +41,7 @@ type DataStoreKey struct {
 	FieldID          string
 }
 
-var _ Key = (*DataStoreKey)(nil)
+var _ Walkable = (*DataStoreKey)(nil)
```
Comment (shahzadlone) [inline internal/keys/datastore_doc.go:44]: question: What inspired the name change, and the context for the name "walkable"? I am guessing it's the key that can be accessed.

EDIT: Was answered by documentation on `Walkable` ignore this question.

File: internal/keys/key.go [hunk 4]
```diff
@@ -20,3 +20,20 @@ type Key interface {
 	Bytes() []byte
 	ToDS() ds.Key
 }
+
+// Walkable represents a key in the database that can be 'walked along'
+// by prefixing the end of the key.
+type Walkable interface {
+	Key
+	PrefixEnd() Walkable
+}
```
Comment (shahzadlone) [inline internal/keys/key.go:29]: praise: Answered my question in another comment, ignore the previous question please.

File: internal/planner/commit.go [hunk 5]
```diff
@@ -142,13 +144,14 @@ func (n *dagScanNode) simpleExplain() (map[string]any, error) {
 
 	// Build the explanation of the spans attribute.
 	spansExplainer := []map[string]any{}
+	undefinedHsKey := keys.HeadStoreKey{}
```
Comment (shahzadlone) [inline internal/planner/commit.go:147]: nitpick:
```suggestion
	undefinedHSKey := keys.HeadStoreKey{}
```

**Review by shahzadlone — APPROVED** [review]
Praise: I like the change, specially the de-abstraction of the `interface` into struct and the removal of `HasValue`

File: internal/db/fetcher/fetcher.go (hunk 2 above)
Comment (AndrewSisley) [reply internal/db/fetcher/fetcher.go:290]: I do not think moving a simple cast into a function would improve the readability of the code here.

File: internal/planner/commit.go (hunk 5 above)
Comment (AndrewSisley) [reply internal/planner/commit.go:147]: Headstore is one word, the current capitalization of `HeadStoreKey` is incorrect and has been fixed in the branchable collections branch (is bundled with some other branchable-specific changes, which is why it isnt in this PR).

File: internal/db/fetcher/fetcher.go (hunk 2 above)
Comment (shahzadlone) [reply internal/db/fetcher/fetcher.go:290]: Can also move `span.Start.(keys.DataStoreKey)` and `span.End.(keys.DataStoreKey)` out of if-else blocks into a variable, and then use that in their respective occurrences. Feel free to ignore as this seems like subjective pref

File: tests/integration/explain/default/delete_test.go (hunk 1 above)
Comment (AndrewSisley) [reply tests/integration/explain/default/delete_test.go:300]: Yeah, there was something fishy going on here with HasValue.  The newly expected result is correct, and the old property/type has died so I didn't bother chasing down exactly where the bug was

**Pull request merged**